### PR TITLE
Powershell as shell

### PR DESCRIPTION
**Pull request by slide**
This allows you to use the ssh-agents-plugin with a Windows system that is running powershell as the default shell (https://docs.microsoft.com/en-us/windows-server/administration/openssh/openssh_server_configuration#configuring-the-default-shell-for-openssh-in-windows). It interrogates the remove system to find out if it is Windows or Linux and then if Powershell is being used. 

There are currently no tests for this, but I am willing to look into tests if that is desired.

### CODE REVIEW

**Comment by slide** [comment]
Looks like the failures are from disk space on the agents, not test failures from these changes.

File: src/main/java/hudson/plugins/sshslaves/SSHLauncher.java [hunk 1]
```diff
@@ -817,18 +851,34 @@ static byte[] readInputStreamIntoByteArrayAndClose(InputStream inputStream) thro
     private void copySlaveJarUsingSCP(TaskListener listener, String workingDirectory) throws IOException, InterruptedException {
         SCPClient scp = new SCPClient(connection);
         try {
-            // check if the working directory exists
-            if (connection.exec("test -d " + workingDirectory ,listener.getLogger())!=0) {
+            if(isWindows) {
+              // check if the working directory exists
+              if (connection.exec("Test-Path " + workingDirectory, listener.getLogger()) != 0) {
```
Comment (KalleOlaviNiemitalo) [inline src/main/java/hudson/plugins/sshslaves/SSHLauncher.java:856]: Should this add quotation marks around `workingDirectory`? The non-Windows `else` branch doesn't do that either, but directory names with spaces in them might be more common on Windows.

File: src/main/java/hudson/plugins/sshslaves/SSHLauncher.java [hunk 2]
```diff
@@ -616,6 +617,34 @@ private void verifyNoHeaderJunk(TaskListener listener) throws IOException, Inter
         }
     }
 
+    private void interrogateRemoteSystem(TaskListener listener) throws IOException, InterruptedException {
```
Comment (kuisathaverat) [inline src/main/java/hudson/plugins/sshslaves/SSHLauncher.java:620]: please, Could you add a javaDoc description?

File: src/main/java/hudson/plugins/sshslaves/SSHLauncher.java [hunk 3]
```diff
@@ -616,6 +617,34 @@ private void verifyNoHeaderJunk(TaskListener listener) throws IOException, Inter
...
+            // now we need to see if the remote shell is cmd.exe or powershell of some sort
+            connection.exec("Write-Host \"ssh-agents-plugin\"",baos);
+            s = fromByteArrayOutputStream(baos);
+            if(s.contains("ssh-agents-plugin")) {
+                isPowershell = true;
+            }
+        }
+    }
+
+    private String fromByteArrayOutputStream(ByteArrayOutputStream baos) throws IOException {
```
Comment (kuisathaverat) [inline src/main/java/hudson/plugins/sshslaves/SSHLauncher.java:638]: please, Could you add a javaDoc description?

File: src/main/java/hudson/plugins/sshslaves/SSHLauncher.java [hunk 4]
```diff
@@ -817,18 +851,34 @@ static byte[] readInputStreamIntoByteArrayAndClose(InputStream inputStream) thro
     private void copySlaveJarUsingSCP(TaskListener listener, String workingDirectory) throws IOException, InterruptedException {
```
Comment (kuisathaverat) [inline src/main/java/hudson/plugins/sshslaves/SSHLauncher.java:851]: please, Could you add a javaDoc description?

File: src/main/java/hudson/plugins/sshslaves/SSHLauncher.java [hunk 5]
```diff
@@ -840,7 +890,11 @@ private void copySlaveJarUsingSCP(TaskListener listener, String workingDirectory
 
     protected void reportEnvironment(TaskListener listener) throws IOException, InterruptedException {
```
Comment (kuisathaverat) [inline src/main/java/hudson/plugins/sshslaves/SSHLauncher.java:891]: please, Could you add a javaDoc description?

File: src/main/java/hudson/plugins/sshslaves/SSHLauncher.java [hunk 6]
```diff
@@ -817,18 +851,34 @@ static byte[] readInputStreamIntoByteArrayAndClose(InputStream inputStream) thro
     private void copySlaveJarUsingSCP(TaskListener listener, String workingDirectory) throws IOException, InterruptedException {
         SCPClient scp = new SCPClient(connection);
         try {
-            // check if the working directory exists
-            if (connection.exec("test -d " + workingDirectory ,listener.getLogger())!=0) {
+            if(isWindows) {
```
Comment (kuisathaverat) [inline src/main/java/hudson/plugins/sshslaves/SSHLauncher.java:854]: Could extract the folder check+creation to a method and write a unit test to cover the new feature?

File: src/main/java/hudson/plugins/sshslaves/SSHLauncher.java (hunk 2 above)
Comment (kuisathaverat) [reply src/main/java/hudson/plugins/sshslaves/SSHLauncher.java:620]: I'd prefer to add two new settings to the Agent configuration, auto detections based on commands outputs use to change with the time and different OS versions could return different results, a setting is always a setting.

File: src/main/java/hudson/plugins/sshslaves/SSHLauncher.java [hunk 7]
```diff
@@ -215,6 +215,11 @@ public class SSHLauncher extends ComputerLauncher {
      */
     private Boolean tcpNoDelay;
 
+
+    private transient boolean isWindows = false;
```
Comment (kuisathaverat) [inline src/main/java/hudson/plugins/sshslaves/SSHLauncher.java:219]: please, Could you add a javaDoc description?

File: src/main/java/hudson/plugins/sshslaves/SSHLauncher.java [hunk 8]
```diff
@@ -215,6 +215,11 @@ public class SSHLauncher extends ComputerLauncher {
...
+
+    private transient boolean isPowershell = false;
```
Comment (kuisathaverat) [inline src/main/java/hudson/plugins/sshslaves/SSHLauncher.java:221]: please, Could you add a javaDoc description?

File: src/main/java/hudson/plugins/sshslaves/SSHLauncher.java [hunk 9]
```diff
@@ -633,6 +662,11 @@ private void startAgent(SlaveComputer computer, final TaskListener listener, Str
         String cmd = "cd \"" + workingDirectory + "\" && " + java + " " + getJvmOptions() + " -jar " + AGENT_JAR +
                      getWorkDirParam(workingDirectory);
 
+        if(isWindows && isPowershell) {
+            cmd = "try { Push-Location \"" + workingDirectory + "\" ; " + java + " " + getJvmOptions() + " -jar " +
+                AGENT_JAR + getWorkDirParam(workingDirectory) + " } finally { Pop-Location }";
+        }
```
Comment (kuisathaverat) [inline src/main/java/hudson/plugins/sshslaves/SSHLauncher.java:668]:
```suggestion
        String cmd = "command no set"
        if(isWindows && isPowershell){
            cmd = "try { Push-Location \"" + workingDirectory + "\" ; " + java + " " + getJvmOptions() + " -jar " +
                AGENT_JAR + getWorkDirParam(workingDirectory) + " } finally { Pop-Location }";
        } else {
            cmd = "cd \"" + workingDirectory + "\" && " + java + " " + getJvmOptions() + " -jar " + AGENT_JAR +
                     getWorkDirParam(workingDirectory);
        }
```

File: src/main/java/hudson/plugins/sshslaves/SSHLauncher.java (hunk 1 above)
Comment (slide) [reply src/main/java/hudson/plugins/sshslaves/SSHLauncher.java:856]: Yes, that is probably a good idea

**Pull request closed without merging**